frontend/internal-packages/agent/src/chat/workflow/nodes/analyzeRequirementsNode.ts

```diff
@@ -1,7 +1,55 @@
+import * as v from 'valibot'
+import { PMAgent } from '../../../langchain/agents'
+import type { BasePromptVariables } from '../../../langchain/utils/types'
+import { convertSchemaToText } from '../../../utils/convertSchemaToText'
 import type { WorkflowState } from '../types'
 
 const NODE_NAME = 'analyzeRequirementsNode'
 
+const requirementsAnalysisSchema = v.object({
+  businessRequirement: v.string(),
+  functionalRequirements: v.record(v.string(), v.array(v.string())),
+  nonFunctionalRequirements: v.record(v.string(), v.array(v.string())),
+})
+
+/**
+ * Execute requirements analysis with retry logic
+ */
+async function executeRequirementsAnalysisWithRetry(
+  pmAgent: PMAgent,
+  promptVariables: BasePromptVariables,
+  logger: WorkflowState['logger'],
+  maxRetries = 3,
+): Promise<v.InferOutput<typeof requirementsAnalysisSchema>> {
+  let lastError: Error | undefined
+
+  for (let attempt = 1; attempt <= maxRetries; attempt++) {
+    try {
+      const response = await pmAgent.analyzeRequirements(promptVariables)
+
+      // Parse and validate JSON response
+      const parsed = JSON.parse(response)
+      return v.parse(requirementsAnalysisSchema, parsed)
```

Comment:
I was not comfortable with the response validation being done here. If we want to use PM agents elsewhere, we need to validate at all callers. Perhaps we should do it this way:

```ts
async analyzeRequirements(variables: BasePromptVariables): Promise<v.inferOutput<typeof requirementsAnalysisSchema>>
```

Also, if `analyzeRequirements` returns only JSON, it would be more effective to use Structured Output.

Reply:
You're absolutely right! I realized that I don't need to `JSON.parse` here; I can simply await the promise within the `analyzeRequirements` function. The changes for the retry logic ended up in the same commit, but I've created the commit as follows: https://github.com/liam-hq/liam/pull/2074/commits/d7ba5c4d76cf786f1ec8629ef5976ac3da16f173